cwms-data-api/src/main/java/cwms/cda/data/dao/project/ProjectLockDao.java

```diff
@@ -1,67 +1,99 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
 package cwms.cda.data.dao.project;
 
 import static cwms.cda.data.dao.project.ProjectDao.toBigInteger;
 
 import cwms.cda.data.dao.JooqDao;
 import cwms.cda.data.dto.project.Lock;
+import cwms.cda.data.dto.project.LockRevokerRights;
 import java.math.BigInteger;
-import java.sql.ResultSet;
-import java.sql.SQLException;
-import java.util.ArrayList;
 import java.util.List;
 import java.util.TimeZone;
 import java.util.logging.Logger;
+import org.jetbrains.annotations.NotNull;
 import org.jooq.DSLContext;
+import org.jooq.Record;
 import org.jooq.SQLDialect;
 import org.jooq.impl.DSL;
 import usace.cwms.db.dao.util.OracleTypeMap;
 import usace.cwms.db.jooq.codegen.packages.CWMS_PROJECT_PACKAGE;
 
 public class ProjectLockDao extends JooqDao<Lock> {
     private static final Logger logger = Logger.getLogger(ProjectLockDao.class.getName());
+    public static final String OFFICE_ID = "OFFICE_ID";
+    public static final String PROJECT_ID = "PROJECT_ID";
+    public static final String APPLICATION_ID = "APPLICATION_ID";
+    public static final String ACQUIRE_TIME = "ACQUIRE_TIME";
+    public static final String SESSION_USER = "SESSION_USER";
+    public static final String OS_USER = "OS_USER";
+    public static final String SESSION_PROGRAM = "SESSION_PROGRAM";
+    public static final String SESSION_MACHINE = "SESSION_MACHINE";
+    public static final String USER_ID = "USER_ID";
 
-    protected ProjectLockDao(DSLContext dsl) {
+    public ProjectLockDao(DSLContext dsl) {
         super(dsl);
     }
 
-    public String requestLock(String projectId, String appId,
-                              boolean revokeExisting, int revokeTimeout, String office) {
+    public String requestLock(String office, String projectId, String appId,
+                              boolean revokeExisting, int revokeTimeout) {
         BigInteger revokeTimeoutBI = toBigInteger(revokeTimeout);
+        String pRevokeExisting = OracleTypeMap.formatBool(revokeExisting);
         return connectionResult(dsl,
                 c -> CWMS_PROJECT_PACKAGE.call_REQUEST_LOCK(getDslContext(c, office).configuration(),
```

Comment:
Need to discuss the usage of this procedure as it internally it grabs session state for the machine, osuser, and program: https://bitbucket.hecdev.net/projects/CWMS/repos/cwms_database/browse/schema/src/cwms/cwms_project_pkg_body.sql#910 to populate the lock data. REGI uses that to identify whether the current user and computer has the project locked: https://bitbucket.hecdev.net/projects/CL/repos/cwms-regi-tools/browse/regi-cache/regi-dao/src/main/java/usace/rowcps/regi/status/AtProjectLockManager.java#378

Reply:
I think you are saying you'd like the CDA user to be able to collect those themselves and pass it into the method call

Reply:
There needs to be some sort of client identification. REGI doesn't persist the project lock id on the client between sessions and there have been times when REGI isn't able to cleanup active locks before a shutdown (process force killed). When the happens, REGI reports a third lock state so that the user is made aware that they have to cleanup their locks. Currently, the client can't just use the JDBC username since CAC authentication uses the CWMS service account user for all access, so osuser and machine name are used instead.

Reply:
The session user will be set to the authenticated user before the SQL call, if that helps.

Reply:
But to be blunt, CDA shouldn't have to care about any distributed transactions... or it it does it needs to be a larger system that works with all the other endpoints (or at least could work time permitting.)

Reply:
Couldn't we duplicate/refactor the REQUEST_LOCK call into one that pulls details out of the session and then calls an inner proc that does the work.  If the CDA call provides all the details CDA could just call the inner call directly.  Then REGI could pass-in whatever it wants for machine-name or os-user.  Those fields got set in the env used by the REGI calls somehow - surely REGI can figure out how to query for them directly and pass them in?

Reply:
That seems like the best solution to me. The way it currently coded, we'd be leaking the CDA server's machine name and osuser back to the client.

Reply:
So the extra fields would be required and CDA only calls the inner proc.   If its just machine-name and os-user that are the problem we could (probably) alternately just set those in the env to the cac-username and the CDA hostname

Reply:
Also, you'd need a `/whoami` endpoint for the client to determine who they are. The user provided apikey doesn't have that information and neither does the X509 certificate.

Also I think you still want the session information to always be correct so that audits in the database are accurate.

Reply:
The apikey doesn't contain the user but the CDA auth is able to map it to a specific db user.  At least that is how I thought it worked.  As long as getting a lock and listing the locks determine the username the same way does it matter if its the os-user or db-user acct?

Reply:
Yeah, the server know's who the client is, but client doesn't know who they are in order to compare against

Reply:
Would it make sense for the client to generate a unique transaction id to track?

Reply:
a UUID type 1, (with the mac address) or version 7 seem the most useful. Or type 3 or 5 to track the user. You wouldn't be able to use it but would be useful to see if one user has consistent usage.

In the case of 7,3, or 5, it would make the most sense to create from CDA, as it has access to the user... and the same clock; which seems pretty important.

Reply:
So basically, client calls a generic `GET transaction` end point to get the UUID and then `DELETE transaction/uuid`.

I am focused on CDA behavior and possible future usage, so not sure how that tracks within the database but doesn't seem like a change that's overly large since you were already encoding information.

Reply:
I don't comprehend your transaction system suggestion.  Maybe it could work.  Would need more than just my input.  I can open an issue for the os-user and extra pieces.

Reply:
https://github.com/USACE/cwms-data-api/issues/711